nfs: add nodeserver

Rakshith-R opened the pull request:
This commit adds nfs nodeserver capable of
mounting nfs volumes, even with pod networking
using NSenter design similar to rbd and cephfs.
NodePublish, NodeUnpublish, NodeGetVolumeStats
and NodeGetCapabilities have been implemented.

Signed-off-by: Rakshith R <rar@redhat.com>

Resolves: #3182

Rakshith-R committed:
nfs: add nodeserver within cephcsi

This commit adds nfs nodeserver capable of
mounting nfs volumes, even with pod networking
using NSenter design similar to rbd and cephfs.
NodePublish, NodeUnpublish, NodeGetVolumeStats
and NodeGetCapabilities have been implemented.

The nodeserver implementation has been inspired
from https://github.com/kubernetes-csi/csi-driver-nfs,
which was previously used for mounted cephcsi exported
nfs volumes. The current implementation is also
backward compatible for the previously created
PVCs.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
deploy: modify nfs daemonset to use cephcsi nfs nodeserver

This commit makes modification to nfs daemonset to use
nfs nodeserver. `nfs.NetNamespaceFilePath` example is
added.

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on deploy/nfs/kubernetes/csi-nfsplugin.yaml:
```diff
@@ -2,154 +2,134 @@
 kind: DaemonSet
 metadata:
-  name: csi-nfs-node
+  name: csi-nfsplugin
```
Document the upgrade scenario as we will have new daemonset and users can delete old daemonsets

Rakshith-R replied:
https://github.com/ceph/ceph-csi/issues/3271

Madhu-1 commented on deploy/nfs/kubernetes/csi-nfsplugin.yaml:
```diff
@@ -2,154 +2,134 @@
+            allowPrivilegeEscalation: true
+          image: registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.5.1
+          args:
+            - "--v=5"
```
set log level to 1?

Madhu-1 commented on internal/nfs/driver/driver.go:
```diff
@@ -43,27 +44,48 @@ func (fs *Driver) Run(conf *util.Config) {
+
+	switch {
+	case conf.IsNodeServer:
+		topology, err := util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
+		if err != nil {
+			log.FatalLogMsg(err.Error())
+		}
```
topology is not supported for nfs just remove it?

Madhu-1 commented on internal/nfs/driver/driver.go:
```diff
@@ -43,27 +44,48 @@ func (fs *Driver) Run(conf *util.Config) {
+	case conf.IsControllerServer:
+		srv.CS = controller.NewControllerServer(cd)
+	case !conf.IsNodeServer && !conf.IsControllerServer:
+		topology, err := util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
```
same here

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,354 @@
+)
+
+const (
+	defaultMountPermission = uint64(0o777)
```
777 is wider permission not possible to get rid of it? we removed 777 for both cephfs and rbd

Rakshith-R replied:
I'll use `0o750` here as default too ?
https://github.com/ceph/ceph-csi/blob/8c5563a9bcef229b6cfe54ae9fbfbbbed70089be/internal/util/util.go#L313-L316

Rakshith-R replied:
```
[1mSTEP[0m: create a PVC and bind it to an app with normal user
Jul 28 11:20:37.978: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{nfs-2540  nfs-2540    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteMany],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-nfs-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,},} to be in Bound state
Jul 28 11:20:37.978: INFO: waiting for PVC nfs-2540 (0 seconds elapsed)
Jul 28 11:20:39.983: INFO: waiting for PVC nfs-2540 (2 seconds elapsed)
Jul 28 11:20:39.990: INFO: Waiting for PV pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 to bind to PVC nfs-2540
Jul 28 11:20:39.990: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [nfs-2540] to have phase Bound
Jul 28 11:20:39.992: INFO: PersistentVolumeClaim nfs-2540 found and phase=Bound (2.726822ms)
Jul 28 11:20:39.992: INFO: Waiting up to 10m0s for PersistentVolume pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 to have phase Bound
Jul 28 11:20:39.995: INFO: PersistentVolume pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 found and phase=Bound (2.674095ms)
Jul 28 11:20:40.009: INFO: Waiting up to pod-run-as-non-root to be in Running state
Jul 28 11:20:56.020: INFO: ExecWithOptions {Command:[/bin/sh -c echo testing > /target/testing] Namespace:nfs-2540 PodName:pod-run-as-non-root ContainerName:write-pod Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Jul 28 11:20:56.020: INFO: >>> kubeConfig: /root/.kube/config
Jul 28 11:20:56.021: INFO: ExecWithOptions: Clientset creation
Jul 28 11:20:56.021: INFO: ExecWithOptions: execute(POST https://192.168.39.123:8443/api/v1/namespaces/nfs-2540/pods/pod-run-as-non-root/exec?command=%2Fbin%2Fsh&command=-c&command=echo+testing+%3E+%2Ftarget%2Ftesting&container=write-pod&container=write-pod&stderr=true&stdout=true)
Jul 28 11:20:56.187: INFO: stdErr occurred: /bin/sh: /target/testing: Permission denied

Jul 28 11:20:56.187: FAIL: failed to validate normal user CephFS pvc and application binding: failed to exec command in pod: command terminated with exit code 1
```

Hey @Madhu-1 , I dont think that will work. 
https://github.com/kubernetes-csi/csi-driver-nfs/blob/c78a98288f3d648cf9d6147fa310444a7f228640/cmd/nfsplugin/main.go#L31
I'll revert to what the csi-nfs-driver used previously 0777. 

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3723/pipeline
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.23/detail/mini-e2e_k8s-1.23/2355/pipeline

Madhu-1 replied:
Do we have securityContext set in the application pod yaml?

Madhu-1 replied:
Looks like we have it. lets target this later.

Rakshith-R replied:
:+1:

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,354 @@
+			if v != "" {
+				mountOptions = append(mountOptions, v)
+			}
+		case mountPermissionsField:
```
do we have this field set by storageclass or the controller?

Rakshith-R replied:
it should be passed through storageclass params, I'll add docs for the same.

nixpanic replied:
are docs missing in the csi-driver-nfs that has this option? Please add a PR for it there as well then.

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,354 @@
+}
+
+// NodePublishVolume mount the volume.
+func (ns *NodeServer) NodePublishVolume(
```
question:- cant we have one NodeStage and multiple NodePublish?

Rakshith-R replied:
> question:- cant we have one NodeStage and multiple NodePublish?

We can but from what i remember from discussion with @nixpanic , 
the kernel remembers nfs mount refs, so if we have one nodePublish for that volume, another nodePublish will just increment the refs and use the same. 
@nixpanic Can you please confirm this.

So there is no particular advantage of having NodeStage and NodePublish. 
and to keep it similar to previously used k8s csi-nfs-driver.

nixpanic replied:
Indeed, NFS should be able to detect that the NFS-server and export are already mounted somewhere, and the in-kernel NFS-client is re-used (and refcounted).

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,354 @@
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
+	}
+	if err != nil {
```
Need to have check for stderr also?

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,357 @@
+
+	if mountPermissions > 0 && !req.GetReadonly() {
+		err = chmodIfPermissionMismatch(ctx, targetPath, os.FileMode(mountPermissions))
+
```
new line not required.

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,357 @@
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
+	switch {
+	case req.GetVolumeId() == "":
+		return fmt.Errorf("volume ID missing in request")
+	case req.GetVolumeCapability() == nil:
+		return fmt.Errorf("volume capability missing in request")
+	case req.GetTargetPath() == "":
+		return fmt.Errorf("target path missing in request")
```
use  errors.New?

Madhu-1 reviewed (commented):
changes LGTM. let's sort out https://github.com/ceph/ceph-csi/pull/3258/files#r930986164 and get this in.

Rakshith-R commented:
@mergifyio rebase

nixpanic commented on internal/nfs/driver/driver.go:
```diff
@@ -43,27 +44,40 @@ func (fs *Driver) Run(conf *util.Config) {
+	switch {
+	case conf.IsNodeServer:
+		srv.NS = nodeserver.NewNodeServer(cd, conf.Vtype)
+
```
why an empty line here?

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,357 @@
+// node server spec.
+type NodeServer struct {
+	*csicommon.DefaultNodeServer
+	Mounter mount.Interface
```
no need to include it if it is already part of the `DefaultNodeServer`?

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,357 @@
+
+	return &NodeServer{
+		DefaultNodeServer: defaultNodeServer,
+		Mounter:           defaultNodeServer.Mounter,
```
I don't think this is needed?

nixpanic reviewed (commented):
How much of the code was taken from the csi-driver-nfs? It would be nice to reference that, and explain that additional options in the StorageClass come from there.

I'd prefer to have a standard (forked?) driver, and then add the nsenter enhancements as a separate commit. Not sure if it makes sense to spend time on that though...

Rakshith-R commented:
> How much of the code was taken from the csi-driver-nfs? It would be nice to reference that, and explain that additional options in the StorageClass come from there.
> 
> I'd prefer to have a standard (forked?) driver, and then add the nsenter enhancements as a separate commit. Not sure if it makes sense to spend time on that though...

Most of the code is rather generic CSI vars,func and procedure.
A few lines are inspired from the csi-driver-nfs to keep it backward compatible
for PVC previously mounted using that driver.
I've referenced the csi-driver-nfs in the commit message.

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e/k8s-1.22

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3758/pipeline
failed to start minikube

Rakshith-R commented:
@mergifyio rebase

Madhu-1 commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,348 @@
+		log.DebugLog(ctx, "performing chmod targetPath(%s, mode:0%o) with permissions(0%o)",
+			targetPath, info.Mode(), mode)
+		if err = os.Chmod(targetPath, mode); err != nil {
+			return err
```
just return err from here if we do so we can remove else check

Rakshith-R replied:
done

Madhu-1 reviewed (dismissed):
changes LGTM

Rakshith-R commented:
hey @nixpanic , Can you please revisit this pr ? 

The ci is failing due to not getting a node to run on .
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3792/pipeline

nixpanic commented on internal/nfs/driver/driver.go:
```diff
@@ -43,27 +44,39 @@ func (fs *Driver) Run(conf *util.Config) {
+		srv.NS = nodeserver.NewNodeServer(cd, conf.Vtype)
+	case conf.IsControllerServer:
+		srv.CS = controller.NewControllerServer(cd)
+	case !conf.IsNodeServer && !conf.IsControllerServer:
```
this can be `default:` instead of a case that covers all other possibilities

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+	source := fmt.Sprintf("%s:%s", server, baseDir)
+
+	err = ns.mountNFS(
+		ctx, volumeID, clusterID, source, targetPath, mountOptions, mountPermissions)
```
very weird linebreak, I wonder why formatting tools don't complain.... or our coding style does not follow the tools?

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+	}
+
+	var server, baseDir, clusterID string
+	for k, v := range req.GetVolumeContext() {
```
some of these parameters are not optional, there needs to be a better verification. Ideally use helper functions for getting the name of the server, the nfs-export (share is Windows/Samba terminology) and others. Return a clear failure when a required option is missing (also easier to add unit-tests for helper functions)

Rakshith-R replied:
This is done valiateNodePublishRequest()

nixpanic replied:
oh, right! That feels a little hidden, I prefer to have a check where the value is actually fetched. Multiple helper functions make it simpler, I think. Not a **must**, but would be _nice_ to have the error checking a little more obvious.

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+	}
+
+	if mountPermissions > 0 && !req.GetReadonly() {
+		err = chmodIfPermissionMismatch(ctx, targetPath, os.FileMode(mountPermissions))
```
is this required, or can it be done by fsGroupPolicy?

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+		return nil, err
+	}
+
+	// Considering kubelet make sure the stage and publish operations
```
kubelet is Kubernetes specific. Is this a requirement of the CSI Spec?

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// NodeUnstageVolume unstage volume.
+func (ns *NodeServer) NodeUnstageVolume(
```
this should not be needed, the DefaultNodeServer should return Unimplemented already.

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+		err                          error
+	)
+
+	if clusterID != "" {
```
anything that calls this function should probably set `clusterID`, don't fetch additional parameters in a mount function.

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
```
add a unit-test for this

nixpanic replied:
btw, Volume in the name of the function is misspelled...

nixpanic replied:
still a typo in the name of the function `validateNodePublishVoluemRequest`

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
```
consider using k8s.io/mount-utils for this, is has lots of error checking and is used elsewhere (RBD) already. Eventually we might be able to extend that package and include the `nsenter` wrapping there too.

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,346 @@
+	return nil
+}
+
+// chmodIfPermissionMismatch only performs chmod when permission mismatches.
```
This isn't recursive, only for a single directory/file. That probably needs explaining. A unit-test would be good to have too. On the other hand, if fsGroupPolicy or similar can cover this, that would be much better.

nixpanic commented on internal/util/csiconfig.go:
```diff
@@ -95,9 +100,9 @@ func readClusterInfo(pathToConfig, clusterID string) (*ClusterInfo, error) {
-	for _, cluster := range config {
-		if cluster.ClusterID == clusterID {
-			return &cluster, nil
+	for i := range config {
```
this looks unrelated to NFS, please remove and create its own PR

Rakshith-R replied:
creating a different commit, linter seems to be catching this error

nixpanic commented on e2e/nfs.go:
```diff
@@ -43,7 +43,7 @@ var (
 	nfsRookCephNFS     = "rook-nfs.yaml"
 	nfsDeploymentName  = "csi-nfsplugin-provisioner"
-	nfsDeamonSetName   = "csi-nfs-node"
+	nfsDeamonSetName   = "csi-nfsplugin"
```
This can be placed in the same commit as where the name is changed, grouping logical changes together. As a rule, each commit should be fully testable, after each commit e2e should be able to run without failures. We don't have that as a strict requirement, but consider that in future PRs. (if there is breakage, `git bisect ..` can not easily be used anymore)

Rakshith-R committed:
cleanup: use index instead of value while iterating

This commit cleans up for loop to use index to access
value instead of copying value into a new variable
while iterating.
```
internal/util/csiconfig.go:103:2: rangeValCopy: each \
iteration copies 136 bytes (consider pointers or indexing) \
(gocritic)
        for _, cluster := range config {
```

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented on internal/csi-common/nodeserver-default.go:
```diff
@@ -29,6 +29,7 @@ import (
 
 // DefaultNodeServer stores driver object.
 type DefaultNodeServer struct {
+	csi.UnimplementedNodeServer
```
now the nodestage and unstage calls will be inherited as unimplemented

Rakshith-R commented:
Removed mountPersmission and chmod, let's keep this first iteration as simple as possible.

added helpers, testcases and addressed reviews.
PTAL

nixpanic commented on internal/nfs/nodeserver/nodeserver.go:
```diff
@@ -0,0 +1,282 @@
+	return nil
+}
+
+// getSource validates volume context, extracts and returns sourceq.
```
`sourceq`? maybe explain what parameters are expected to be set in the map?

Rakshith-R replied:
done

nixpanic commented on internal/nfs/nodeserver/nodeserver_test.go:
```diff
@@ -0,0 +1,163 @@
+		wantErr bool
+	}{
+		{
+			name: "ipv4 address",
```
this does not need to be an IP-address, it can be a hostname too. Please add a test-case for that.

Rakshith-R replied:
done

nixpanic reviewed (changes requested):
some minor modifications needed, looks pretty complete to me now. Thanks!

nixpanic commented on internal/nfs/nodeserver/nodeserver_test.go:
```diff
@@ -0,0 +1,174 @@
+			name: "hostname as address",
+			args: args{
+				volContext: map[string]string{
+					paramServer: "hostname.io",
```
nit: use `example.io` or other official `example.*` hostnames.

Rakshith-R replied:
done :)

nixpanic reviewed (changes requested):
_so close!_ just a few nits left.

Rakshith-R committed:
deploy: update fsGroupPolicy to File for nfs driver object

NFS volume similar to CephFs volumes can support
fsGroupPolicy as File, now Kubernetes Kubernetes may
use fsGroup to change permissions and ownership of the
volume to match user requested fsGroup in the pod's
SecurityPolicy regardless of fstype or access mode.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
@mergifyio rebase 

added a commit to use fsgroupPolicy as File for nfs similar to cephfs, 
ptal

Madhu-1 commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio rebase
